File: src-cpp/CarController.cpp
#include "CarController.hpp"
#include "Scheduler.hpp"
#include <algorithm>

// 构造函数，初始化小车控制器
CarController::CarController(int id, std::map<int, PortDevice *> &deviceMap, double initialPosition)
    : id(id), deviceMap(deviceMap), position(initialPosition)
{
}

// 设置调度器
void CarController::setScheduler(Scheduler *scheduler)
{
    this->scheduler = scheduler;
}

// 分配任务，设置起止点
void CarController::assignTask(std::shared_ptr<CarTask> task, double fromPos, double toPos)
{
    this->task = task;
    this->portFrom = deviceMap[task->fromDevice];
    this->portTo = deviceMap[task->toDevice];
    this->status = CarStatus::Moving;
    stopPoints = {fromPos, toPos};
    currentStopIndex = 0;
    hasMaterial = false;
    setTarget(stopPoints[0]);
}

// 尝试恢复运行
void CarController::tryResume()
{
    if (status == CarStatus::Waiting)
    {
        setTargetSpeed(maxStraightSpeed);
        status = CarStatus::Moving;
    }
}

// 设置目标位置
void CarController::setTarget(double pos)
{
    // 简单逻辑；更复杂的逻辑可能涉及路径规划
    if (pos < position)
        targetSpeed = -maxStraightSpeed;
    else
        targetSpeed = maxStraightSpeed;
}

// 设置目标速度
void CarController::setTargetSpeed(double v)
{
    targetSpeed = v;
}

// 更新小车状态
void CarController::update(double dt)
{
    if (status == CarStatus::Idle || isCollision)
        return;

    updateMotion(dt);

    // 到达目标点后的处理
    if (status == CarStatus::Moving && reachedTarget())
    {
        if (currentStopIndex == 0)
        {
            status = CarStatus::Loading; // 装载中
            loadingTimer = 0.0;
        }
        else if (currentStopIndex == 1)
        {
            status = CarStatus::Unloading; // 卸载中
            unloadingTimer = 0.0;
        }
        setTargetSpeed(0);
    }

    // 装载过程
    if (status == CarStatus::Loading)
    {
        loadingTimer += dt;
        if (loadingTimer >= loadingDuration)
        {
            status = CarStatus::Loaded;
            hasMaterial = true;
            currentStopIndex++;
            setTarget(stopPoints[currentStopIndex]);
            status = CarStatus::Moving;
        }
    }

    // 卸载过程
    if (status == CarStatus::Unloading)
    {
        unloadingTimer += dt;
        if (unloadingTimer >= unloadingDuration)
        {
            status = CarStatus::Idle;
            hasMaterial = false;
            task = nullptr;
        }
    }

    // 记录速度表
    if (addCarSpeedTable)
    {
        addCarSpeedTable(id, position, speed, static_cast<int>(status));
    }
}

// 更新小车运动状态
void CarController::updateMotion(double dt)
{
    double maxSpeed = inCurve(position) ? maxCurveSpeed : maxStraightSpeed;
    if (std::abs(targetSpeed) > maxSpeed)
        targetSpeed = (targetSpeed > 0 ? 1 : -1) * maxSpeed;

    acceleration = (targetSpeed - speed) * 2.0;
    speed += acceleration * dt;
    position += speed * dt;

    // 环形轨道处理
    if (position >= trackLength)
        position -= trackLength;
    else if (position < 0)
        position += trackLength;
}

// 判断当前位置是否在弯道
bool CarController::inCurve(double pos) const
{
    for (const auto &range : curveRanges)
    {
        if (pos >= range.first && pos <= range.second)
            return true;
    }
    return false;
}

// 判断是否到达目标点
bool CarController::reachedTarget() const
{
    return std::abs(position - stopPoints[currentStopIndex]) < 0.5;
}

// 获取当前位置
double CarController::getPosition() const
{
    return position;
}

// 获取当前速度
double CarController::getSpeed() const
{
    return speed;
}

// 获取当前状态
CarStatus CarController::getStatus() const
{
    return status;
}

// 获取与另一辆车的距离
double CarController::getDistanceTo(const CarController &car, double trackLen) const
{
    double dist = car.position - position;
    if (dist < 0)
        dist += trackLen;
    return dist;
}

// 获取最大直线速度
double CarController::getMaxStraightSpeed() const
{
    return maxStraightSpeed;
}

// 获取显示信息
CarController::DisplayInfo CarController::getDisplayProps() const
{
    return DisplayInfo{position, status};
}

// 获取全部信息
CarController::FullInfo CarController::getAllInfo() const
{
    return FullInfo{
        id, position, speed, acceleration, status, task, portFrom, portTo, hasMaterial};
}

File: src/utils/scheduler1.0/CarController.ts
import { invoke } from '@tauri-apps/api/tauri'
import type { CarTask } from '@/types'
import type { PortDevice } from './PortDevice.ts'
import store from '@/store'

type CarStatus = 'idle' | 'moving' | 'loading' | 'loaded' | 'unloading' | 'waiting' | 'cruising'

export class CarController {
  public id: number
  public deviceMap: Map<number, PortDevice>
  constructor(id: number, deviceMap: Map<number, PortDevice>, initialPosition = 0) {
    this.id = id
    this.deviceMap = deviceMap
  }

  public async update(dt: number) {
    // 让 Rust 后端处理 update，返回最新状态
    const info = await invoke('car_update', { id: this.id, dt })
    return info
  }

  public async assignTask(task: CarTask, fromPos: number, toPos: number) {
    return await invoke('car_assign_task', { id: this.id, task, fromPos, toPos })
  }

  public async setTarget(pos: number) {
    return await invoke('car_set_target', { id: this.id, pos })
  }

  public async setTargetSpeed(v: number) {
    return await invoke('car_set_target_speed', { id: this.id, v })
  }

  public async tryResume() {
    return await invoke('car_try_resume', { id: this.id })
  }

  public async getDisplayProps() {
    return await invoke('car_get_display_props', { id: this.id })
  }

  public async getPosition() {
    return await invoke('car_get_position', { id: this.id })
  }

  public async getSpeed() {
    return await invoke('car_get_speed', { id: this.id })
  }

  public async getStatus() {
    return await invoke('car_get_status', { id: this.id })
  }

  public async getAllInfo() {
    return await invoke('get_car_info', { id: this.id })
  }

  public async getDistanceTo(car: CarController, trackLength: number) {
    return await invoke('car_get_distance_to', { id: this.id, otherId: car.id, trackLength })
  }

  public async getMaxStraightSpeed() {
    return await invoke('car_get_max_straight_speed', { id: this.id })
  }
}
